### PR TITLE
Deduplicate execution payload envelopes on gossip

### PR DESCRIPTION
**Pull request by conache**
## Issue Addressed

Part of #8948. 

This PR implements the envelope deduplication check, needed as part of the gossip verification of received payload envelopes (per Gloas [consensus-specs](https://github.com/ethereum/consensus-specs/blob/master/specs/gloas/p2p-interface.md#new-execution_payload)):

> [IGNORE] The node has not seen another valid envelope for this block root from this builder.

## Proposed Changes

- Add a new cache which tracks if a valid envelope was already received from `builder_index` for `block_root`
- The cache is checked and updated as part of `GossipVerifiedEnvelope::new`. In case a valid envelope was already observed for the `(block_root, builder_index)` pair, the function returns `EnvelopeError::EnvelopeAlreadySeen`, which maps to `IGNORE`
- Seen envelopes are pruned per-slot, alongside the existing gossip caches

### CODE REVIEW

File: beacon_node/beacon_chain/src/payload_envelope_verification/gossip_seen_envelope_cache.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+impl GossipSeenEnvelopeCache {
+    /// Mark the payload envelope as seen for the tuple `(slot, block_root, builder_index)`
+    pub fn mark_envelope_seen(&self, slot: Slot, block_root: Hash256, builder_index: BuilderIndex) {
+        let mut seen_envelopes = self.seen_envelopes.write();
+        seen_envelopes
+            .entry(slot)
+            .or_default()
+            .insert((block_root, builder_index));
+    }
```
Comment (eserilev) [inline beacon_node/beacon_chain/src/payload_envelope_verification/gossip_seen_envelope_cache.rs]: i would rather pass in the full `GossipVerifiedEnvelope` here for type safety guarantees that the envelope passed into this function has passed gossip verification

File: beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs [hunk 2]
```diff
@@ -263,6 +277,13 @@ impl<T: BeaconChainTypes> GossipVerifiedEnvelope<T> {
             return Err(EnvelopeError::BadSignature);
         }
 
+        // Mark gossip-validated envelope as seen for the (block_root, builder_index) pair
+        ctx.gossip_seen_envelope_cache.mark_envelope_seen(
+            block_slot,
+            beacon_block_root,
+            builder_index,
```
Comment (eserilev) [inline beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs]: we should make `mark_envelope_seen` atomic. right now theres a race condition where two envelopes arriving at the same time pass the above `has_seen_envelope` check. we would then forward both envelopes (instead of ignoring one of them). not really a big deal but we can close this race if we have `mark_envelope_seen` return if the envelope has already been seen or not

File: beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs [hunk 3]
```diff
@@ -161,6 +165,18 @@ impl<T: BeaconChainTypes> GossipVerifiedEnvelope<T> {
...
+        if ctx.gossip_seen_envelope_cache.has_seen_envelope(
+            block_slot,
+            beacon_block_root,
+            builder_index,
+        ) {
+            return Err(EnvelopeError::EnvelopeAlreadySeen {
+                block_root: beacon_block_root,
+                builder_index,
+            });
+        }
```
Comment (eserilev) [inline beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs]: verify_envelope_for_gossip is also used for publishing enveloeps over gossip. if some failure causes a VC/builder to republish its own envelope, we end up raising an error on the http path. We might want to catch that error in the http path and return an `Ok` instead since the envelope was already verified and broadcasted. 

Not 100% sure how we want to handle that case tbh

File: beacon_node/beacon_chain/src/payload_envelope_verification/gossip_seen_envelope_cache.rs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        builder_index: BuilderIndex,
+    ) -> bool {
+        self.seen_envelopes
+            .read()
+            .get(&slot)
+            .is_some_and(|seen| seen.contains(&(block_root, builder_index)))
+    }
+
+    /// Prune anything before `current_slot`
+    pub fn prune(&self, current_slot: Slot) {
```
Comment (eserilev) [inline beacon_node/beacon_chain/src/payload_envelope_verification/gossip_seen_envelope_cache.rs]: we are pruning too aggressively. there is no rule in gossip that states the envelope must be for the current slot. the only rules are that its not prior to the finalized slot. i think we can retain all the way up to the finalized slot OR a flat 64 slot window like the `envelope_times_cache`

File: beacon_node/beacon_chain/src/payload_envelope_verification/gossip_seen_envelope_cache.rs (hunk 1 above)
Comment (conache) [reply beacon_node/beacon_chain/src/payload_envelope_verification/gossip_seen_envelope_cache.rs]: Addressed in https://github.com/sigp/lighthouse/pull/9775/changes/0ad58552df98ca01a610abba64d1a4e9bcf09ff8

File: beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs (hunk 2 above)
Comment (conache) [reply beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs]: got it, thanks! 🙏 This should be fixed in https://github.com/sigp/lighthouse/pull/9775/changes/0ad58552df98ca01a610abba64d1a4e9bcf09ff8

File: beacon_node/beacon_chain/src/payload_envelope_verification/gossip_seen_envelope_cache.rs (hunk 4 above)
Comment (conache) [reply beacon_node/beacon_chain/src/payload_envelope_verification/gossip_seen_envelope_cache.rs]: we should only prune slots prior the finalized one now. Added the change in https://github.com/sigp/lighthouse/pull/9775/commits/f0205228936bcbf384b36e1c73851468988ec7e4

File: beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs (hunk 3 above)
Comment (conache) [reply beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs]: Pushed your suggestion in https://github.com/sigp/lighthouse/pull/9775/changes/bf2b82122e3f81bfc02593bc7e7fcce2b09a83df. We catch the `EnvelopeAlreadySeen` error and should return `Ok`  now.

However, I'm a bit concerned that returning a success response early in the second call wouldn't guarantee that the envelope was actually broadcasted in the previous one. This could happen if the first call breaks before broadcasting ([here](https://github.com/sigp/lighthouse/blob/bf2b82122e3f81bfc02593bc7e7fcce2b09a83df/beacon_node/http_api/src/beacon/execution_payload_envelopes.rs#L272-L286)) or if the broadcast itself didn't work for a reason.
    
I'm not sure what's the best way to handle this, but I was thinking that we could add an argument like `IgnoreDuplicates::{Yes, No}` to `verify_envelope_for_gossip` that would help "disable" the dedup on the http path and only keep it enabled on the gossip path.
    
Curious to know what you think on this

Comment (eserilev) [reply beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs]: yes i think for the http path only we should just effectively ignore the cache and always gossip verify and broadcast. That `IgnoreDuplicates` enum sounds like a good way to do that

Comment (conache) [reply beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs]: sounds good! Just added this mechanism, with the only note that I ended up using a different name for it - `AllowDuplicates`

**Comment by eserilev** [comment]
Been  thinking about this one a bit more and I think we should always pass `AllowDuplicates::Yes` on the RPC path. the reason being is that if we mark the envelope as seen in the cache, but for some reason we fail to write it (some transient EL error for example) we'll never be able to lookup this envelope again. That could get our node stuck in an unrecoverable state. 

Since the RPC path never re-gossips, I think allowing duplicates is fine

I think this also means we can get rid of `AllowDuplicates` and use `EnvelopeSource` enum instead. I think you'd have to move it out of the metrics crate and add an `Http` variant to it though

File: beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs
Comment (eserilev) [inline beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs]: I think this PR allows us to delete this TODO

Comment (conache) [reply beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs]: oh, definitely! removed in https://github.com/sigp/lighthouse/pull/9775/changes/bf04b6652da1e1c7ca6a9d5bda66fc263f9dfb51

**Comment by conache** [comment]
> Been thinking about this one a bit more and I think we should always pass `AllowDuplicates::Yes` on the RPC path. the reason being is that if we mark the envelope as seen in the cache, but for some reason we fail to write it (some transient EL error for example) we'll never be able to lookup this envelope again. That could get our node stuck in an unrecoverable state.
> 
> Since the RPC path never re-gossips, I think allowing duplicates is fine
> 
> I think this also means we can get rid of `AllowDuplicates` and use `EnvelopeSource` enum instead. I think you'd have to move it out of the metrics crate and add an `Http` variant to it though

Thanks for flagging the RPC case, it makes sense! It's also great that we can already use an existing enum for envelopes from different paths. I went ahead and removed `AllowDuplicates`. We're now using `EnvelopeSource`, which I moved to the `payload_envelope_verification` module in `beacon_chain`.

Envelopes are marked as seen on every path, but deduplication is now only enforced when the envelope was received over gossip. I also added more tests covering duplicates received on each possible path.

I also went ahead and renamed the `GossipSeenEnvelopeCache` to `ObservedPayloadEnvelopes`. My initial thinking was that this cache would be used only on the gossip path, but since envelopes are now marked on every path, the renaming made sense to me.

Everything should be ready for another look 🙏

File: beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs [hunk 5]
```diff
@@ -141,6 +144,8 @@ impl<T: BeaconChainTypes> GossipVerifiedEnvelope<T> {
     ) -> Result<Self, EnvelopeError> {
         let envelope = &signed_envelope.message;
         let beacon_block_root = envelope.beacon_block_root;
+        let block_slot = envelope.slot();
```
Comment (eserilev) [inline beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs]: a maliciously replayed envelope with a mutated slot could cost us a full bls verification that we could have avoided. We already fetch the block from fork choice below, so we can just use `proto_block.slot` instead

File: beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs [hunk 6]
```diff
@@ -263,24 +279,45 @@ impl<T: BeaconChainTypes> GossipVerifiedEnvelope<T> {
...
+                    slot: block_slot,
                     builder_index,
-                    block_hash: signed_envelope.message.payload.block_hash,
+                    block_hash: gossip_verified_envelope
+                        .signed_envelope
+                        .message
+                        .payload
+                        .block_hash,
                     block_root: beacon_block_root,
                 },
```
Comment (eserilev) [inline beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs]: we are potentially re-emitting SSE events for envelopes we've already seen in the Http/Rpc case. If the envelope has already been seen we should skip emitting the SSE event

**Review by eserilev — CHANGES_REQUESTED** [review]
a few more minor changes and then we will merge. thanks for sticking with this!

File: beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs (hunk 5 above)
Comment (conache) [reply beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs]: this makes sense, thanks for the explanation! 🙏  I switched to using the `proto_block.slot` in https://github.com/sigp/lighthouse/pull/9775/commits/513cf24de9700ac8c6b6fc9aa5b66d5fb16c764e and added the test for this case in https://github.com/sigp/lighthouse/pull/9775/changes/6e90f63819383248e4bfdcd532613eb7ca84c023

File: beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs (hunk 6 above)
Comment (conache) [reply beacon_node/beacon_chain/src/payload_envelope_verification/gossip_verified_envelope.rs]: fixed and added a small comment in https://github.com/sigp/lighthouse/pull/9775/commits/6480c8055daa79fbba4909306d6d4c245bf1180c

File: beacon_node/beacon_chain/src/payload_envelope_verification/observed_payload_envelopes.rs [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    pub fn observe_envelope<T: BeaconChainTypes>(
+        &self,
+        envelope: &GossipVerifiedEnvelope<T>,
+    ) -> bool {
+        let message = &envelope.signed_envelope.message;
+        self.seen_envelopes
+            .write()
+            .entry(message.slot())
+            .or_default()
+            .insert((message.beacon_block_root, message.builder_index))
```
Comment (eserilev) [inline beacon_node/beacon_chain/src/payload_envelope_verification/observed_payload_envelopes.rs:34]: one last nit

Instead of using `message.slot` (which comes from gossip) we should use `envelope.block.slot` which comes from the store

Comment (conache) [reply beacon_node/beacon_chain/src/payload_envelope_verification/observed_payload_envelopes.rs:34]: Got it, this should be fixed now

**Review by eserilev — APPROVED** [review]
LGTM, thanks!
